mono/metadata/file-mmap-windows.c

```diff
@@ -1,71 +1,413 @@
+	FILE_ALREADY_EXISTS,
+	PATH_TOO_LONG,
+	COULD_NOT_OPEN,
+	CAPACITY_MUST_BE_POSITIVE,
+	INVALID_FILE_MODE,
+	COULD_NOT_MAP_MEMORY,
+	ACCESS_DENIED,
+	CAPACITY_LARGER_THAN_LOGICAL_ADDRESS_SPACE
+};
+
+enum {
+	FILE_MODE_CREATE_NEW = 1,
+	FILE_MODE_CREATE = 2,
+	FILE_MODE_OPEN = 3,
+	FILE_MODE_OPEN_OR_CREATE = 4,
+	FILE_MODE_TRUNCATE = 5,
+	FILE_MODE_APPEND = 6,
+};
+
+enum {
+	MMAP_FILE_ACCESS_READ_WRITE = 0,
+	MMAP_FILE_ACCESS_READ = 1,
+	MMAP_FILE_ACCESS_WRITE = 2,
+	MMAP_FILE_ACCESS_COPY_ON_WRITE = 3,
+	MMAP_FILE_ACCESS_READ_EXECUTE = 4,
+	MMAP_FILE_ACCESS_READ_WRITE_EXECUTE = 5,
+};
+
+static DWORD get_page_access (int access)
 {
-	g_error ("No windows backend");
-	return NULL;
+	switch (access) {
+	case MMAP_FILE_ACCESS_READ:
+		return PAGE_READONLY;
+	case MMAP_FILE_ACCESS_READ_WRITE:
+		return PAGE_READWRITE;
+	case MMAP_FILE_ACCESS_COPY_ON_WRITE:
+		return PAGE_WRITECOPY;
+	case MMAP_FILE_ACCESS_READ_EXECUTE:
+		return PAGE_EXECUTE_READ;
+	case MMAP_FILE_ACCESS_READ_WRITE_EXECUTE:
+		return PAGE_EXECUTE_READWRITE;
+	default:
+		g_error ("unknown MemoryMappedFileAccess %d", access);
+	}
 }
 
-void *
-mono_mmap_open_handle (void *handle, MonoString *mapName, gint64 *capacity, int access, int options, int *error)
+static DWORD get_file_access (int access)
 {
-	g_error ("No windows backend");
-	return NULL;
+	switch (access) {
+	case MMAP_FILE_ACCESS_READ:
+	case MMAP_FILE_ACCESS_READ_EXECUTE:
+		return GENERIC_READ;
+	case MMAP_FILE_ACCESS_READ_WRITE:
+	case MMAP_FILE_ACCESS_COPY_ON_WRITE:
+	case MMAP_FILE_ACCESS_READ_WRITE_EXECUTE:
+		return GENERIC_READ | GENERIC_WRITE;
+	case MMAP_FILE_ACCESS_WRITE:
+		return GENERIC_WRITE;
+	default:
+		g_error ("unknown MemoryMappedFileAccess %d", access);
+	}
 }
 
-void
-mono_mmap_close (void *mmap_handle)
+static int get_file_map_access (int access)
 {
-	g_error ("No windows backend");
+	switch (access) {
+	case MMAP_FILE_ACCESS_READ:
+		return FILE_MAP_READ;
+	case MMAP_FILE_ACCESS_WRITE:
+		return FILE_MAP_WRITE;
+	case MMAP_FILE_ACCESS_READ_WRITE:
+		return FILE_MAP_READ | FILE_MAP_WRITE;
+	case MMAP_FILE_ACCESS_COPY_ON_WRITE:
+		return FILE_MAP_COPY;
+	case MMAP_FILE_ACCESS_READ_EXECUTE:
+		return FILE_MAP_EXECUTE | FILE_MAP_READ;
+	case MMAP_FILE_ACCESS_READ_WRITE_EXECUTE:
+		return FILE_MAP_EXECUTE | FILE_MAP_READ | FILE_MAP_WRITE;
+	default:
+		g_error ("unknown MemoryMappedFileAccess %d", access);
+	}
 }
 
-void
-mono_mmap_configure_inheritability (void *mmap_handle, gboolean inheritability)
+static int convert_win32_error (int error, int def)
 {
-	g_error ("No windows backend");
+	switch (error) {
+	case ERROR_FILE_NOT_FOUND:
+		return FILE_NOT_FOUND;
+	case ERROR_FILE_EXISTS:
+	case ERROR_ALREADY_EXISTS:
+		return FILE_ALREADY_EXISTS;
+	case ERROR_ACCESS_DENIED:
+		return ACCESS_DENIED;
+	}
+	return def;
 }
 
-void
-mono_mmap_flush (void *mmap_handle)
+static void *open_handle (void *handle, MonoString *mapName, int mode, gint64 *capacity, int access, int options, int *error)
```

Comment:
These `int *error` could be replaced with `IOError *error` or something like that, if you named the `enum`.